benchmark/get_cath.py

```diff
@@ -216,23 +251,761 @@ def get_pdbs(
+        .groupby(by="sequence")
+        .mean()
+    )
+    prec, rec, f1, sup = metrics.precision_recall_fscore_support(sequence, prediction)
+
+    entropy_frame.loc[:, "recall"] = rec
+    entropy_frame.loc[:, "precision"] = prec
+    entropy_frame.loc[:, "f1"] = f1
+    entropy_frame.loc[:, "auc"] = roc_auc
+    return entropy_frame
+
+
+def get_angles(protein: pd.Series, path_to_assemblies: Path) -> np.array:
+    """Gets backbone torsion angles for protein.
+
+    Parameters
+    ----------
+        protein: pd.Series
+            Series containing protein info.
+        path_to_assemblies: Path
+            Path to the directory with biological assemblies.
+    Returns
+    -------
+    torsion_angles: np.array
+        Array with torsion angles."""
+
+    path = path_to_assemblies / protein.PDB[1:3] / f"pdb{protein.PDB}.ent.gz"
+    if path.exists():
+        with gzip.open(path, "rb") as file:
+            assembly = ampal.load_pdb(file.read().decode(), path=False)
+            # check is assembly has multiple states, pick the first
+            if isinstance(assembly, ampal.AmpalContainer):
+                assembly = assembly[0]
+            chain = assembly[protein.chain]
+            torsion_angles = ampal.analyse_protein.measure_torsion_angles(chain)
+    return torsion_angles
+
+
+def format_angle_sequence(
+    df: pd.DataFrame,
+    predictions: dict,
+    path_to_assemblies: Path,
+    by_fragment: bool = False,
+    ignore_uncommon=False,
+) -> Tuple[str, Iterable, str, List[List[float]]]:
+    """Gets Psi and Phi angles for all residues in predictions, can skip uncommon acids.
+
+
+    Parameters
+    ----------
+    df: pd.DataFrame
+        DataFrame with CATH fragment info. The frame must have predicted sequence, true sequence and start/stop index of CATH fragment.
+    predictions: dict
+        Dictionary with loaded predictions.
+    path_to_assemblies: Path
+        Path to the directory with biological assemblies.
+    by_fragment: bool
+        If true scores only CATH fragments, if False, scores entire chain.
+    ignore_uncommon=True
+        If True, ignores uncommon residues in accuracy calculations.
+        
+    Returns
+    -------
+    sequece:str
+        Protein sequence.
+    prediction: str or np.array
+        Predicted protein sequence or probability matrix.
+    dssp: str
+        String with dssp data
+    torsion:List[List[float]]
+        List with torsion angles. Format:[[omega,phi,psi]].
+    """
+
+    sequence = ""
+    dssp = ""
+    torsion = []
+    prediction = np.empty([0, 20])
+    for i, protein in df.iterrows():
+        if protein.PDB + protein.chain in predictions:
+            start = protein.start
+            stop = protein.stop
+            predicted_sequence = predictions[protein.PDB + protein.chain]
+            protein_angle = get_angles(protein, path_to_assemblies)
+
+            # remove uncommon acids
+            if ignore_uncommon and protein.uncommon_index != []:
```

Comment:
Be careful with more complex logical statements, it's always worth wrapping wrapping up component tests in brackets.